test/modules/aggregation-layers/heatmap-layer/heatmap-layer.spec.js

```diff
@@ -0,0 +1,220 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+import test from 'tape-catch';
+import * as FIXTURES from 'deck.gl-test/data';
+import {testLayer, testInitializeLayer, generateLayerTests} from '@deck.gl/test-utils';
+import {MapView} from '@deck.gl/core';
+import {HeatmapLayer} from '@deck.gl/aggregation-layers';
+import {default as TriangleLayer} from '@deck.gl/aggregation-layers/heatmap-layer/triangle-layer';
+
+const getPosition = d => d.COORDINATES;
+
+const viewport1 = new MapView().makeViewport({
+  width: 100,
+  height: 100,
+  viewState: {longitude: 0, latitude: 10, zoom: 10}
+});
+
+const viewport2_slightChange = new MapView().makeViewport({
+  width: 100,
+  height: 100,
+  viewState: {longitude: 0.01, latitude: 10, zoom: 10}
+});
+
+const viewport3_bigChange = new MapView().makeViewport({
+  width: 100,
+  height: 100,
+  viewState: {longitude: 10, latitude: 0, zoom: 10}
+});
+
+const viewport4_zoomChange = new MapView().makeViewport({
+  width: 100,
+  height: 100,
+  viewState: {longitude: 10, latitude: 0, zoom: 9}
+});
+
+test('HeatmapLayer', t => {
+  const testCases = generateLayerTests({
+    Layer: HeatmapLayer,
+    sampleProps: {
+      data: FIXTURES.points.slice(0, 3),
+      getPosition
+    },
+    assert: t.ok,
+    onBeforeUpdate: ({testCase}) => t.comment(testCase.title),
+    onAfterUpdate({layer}) {
+      t.ok(layer.state.worldBounds, 'should update state.worldBounds');
+    }
+  });
+
+  testLayer({Layer: HeatmapLayer, testCases, onError: t.notOk});
+
+  t.end();
+});
+
+test('HeatmapLayer#renderSubLayer', t => {
+  const layer = new HeatmapLayer({
+    id: 'contourLayer',
+    data: FIXTURES.points,
+    getPosition
+  });
+
+  testInitializeLayer({layer, onError: t.notOk});
+
+  // render sublayer
+  const subLayer = layer.renderLayers();
+  testInitializeLayer({layer: subLayer, onError: t.notOk});
+
+  t.ok(subLayer instanceof TriangleLayer, 'Sublayer Triangle layer rendered');
+
+  t.end();
```

Comment:
`testInitializeLayer` does not handle finalizing a layer. Either call `layer._finalize()` or merge this into the next test.